### PR TITLE
Improve generator performance

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

The _Determine ARM resource relationships_ stage was taking around a minute to run on my laptop, so I took a look at it - and discovered it's performance was O(n<sup>2</sup>); I modified it to run in O(n) time.

Performance change on my laptop:

| Before  | After  |
| -------:| ------:|
| 1m9.73s | 6.378s |
|  52.48s | 3.014s |

Performance change in CI:

| Before  | After  |
| -------:| ------:|
| 15.258s | 1.333s |
| 15.239s |        |

**Special notes for your reviewer**:

The `findChildren()` method is no longer strictly necessary, but retaining it meant I could keep the tests largely intact.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

**Comment by theunrepentantgeek** [comment]
_PR Summary by CoPilot ..._

This pull request modifies the `determine_resource_ownership.go` and `determine_resource_ownership_test.go` files in the `v2/tools/generator/internal/codegen/pipeline` directory. The changes involve refactoring the `DetermineResourceOwnership` function and related functions to improve the code's readability, maintainability, and error handling.

The most important changes include:

Refactoring of `DetermineResourceOwnership` function:

* [`v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go`](diffhunk://#diff-674b639cf91d1c9d7c006d2033cd0c48cf3644de0a02e792a76ec8a542100bd5L26-R94): The `DetermineResourceOwnership` function was refactored to use a new `ownershipStage` struct, which encapsulates the state and behavior related to determining resource ownership. The function now creates an instance of `ownershipStage` and uses its methods to perform the ownership determination.

Introduction of `ownershipStage` struct:

* [`v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go`](diffhunk://#diff-674b639cf91d1c9d7c006d2033cd0c48cf3644de0a02e792a76ec8a542100bd5L26-R94): The new `ownershipStage` struct and its methods were introduced. These methods encapsulate the logic for determining resource ownership, indexing resources by their parent, and finding and assigning owners to child resources. [[1]](diffhunk://#diff-674b639cf91d1c9d7c006d2033cd0c48cf3644de0a02e792a76ec8a542100bd5L26-R94) [[2]](diffhunk://#diff-674b639cf91d1c9d7c006d2033cd0c48cf3644de0a02e792a76ec8a542100bd5L61-R146)

Improvements to error handling:

* [`v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go`](diffhunk://#diff-674b639cf91d1c9d7c006d2033cd0c48cf3644de0a02e792a76ec8a542100bd5L61-R146): Improved error handling was implemented in the `assignOwners` method of the `ownershipStage` struct. Now, instead of returning immediately when an error occurs, the function collects all errors and returns them as an aggregate error at the end.

Refactoring of test cases:

* [`v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership_test.go`](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R24-R34): The test cases were refactored to use the new `ownershipStage` struct and its methods. This makes the test cases more consistent with the changes to the code they are testing. [[1]](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R24-R34) [[2]](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R43-R57) [[3]](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R66-R80) [[4]](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R89-R103) [[5]](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R112-R126) [[6]](diffhunk://#diff-a38de4d013b423f90de65e45b5b603dc36148e5349ed1efb6c8d4f413ed5f622R135-R149)

Minor Bug Fix:

* [`v2/tools/generator/internal/codegen/pipeline/resource_registration_file.go`](diffhunk://#diff-2246e58762511defabcdcc55cc60d997b91d1d8621a939065df2825a26c9011bL104-L106): Removed an unnecessary error check that was causing a compile error.

File: v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go [hunk 1]
```diff
@@ -58,75 +100,50 @@ func determineOwnership(
 		updatedDefs[resolved.SpecDef.Name()] = newDef
 	}
 
-	setDefaultOwner(configuration, definitions, updatedDefs)
+	if len(errs) > 0 {
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go:103]: extremely minor:
Could also be:
```
err := kerrors.NewAggregate(errs)
if err != nil {
     errors.Wrapf(
			err,
			"failed to update ownership for some resources")
}
```

which I think is the pattern we usually use.

File: v2/tools/generator/internal/codegen/pipeline/resource_registration_file.go [hunk 2]
```diff
@@ -101,9 +101,6 @@ func (r *ResourceRegistrationFile) AsAst() (*dst.File, error) {
 
 	// Create Resource Extensions
 	resourceExtensionTypes := r.createGetResourceExtensions(codeGenContext)
-	if err != nil {
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/pipeline/resource_registration_file.go:104]: How did you notice this?

**Review by matthchr — APPROVED** [review]

File: v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/codegen/pipeline/determine_resource_ownership.go:103]: Using `len(errs) > 0` is the usual way I've done it - lots of matches to `git grep "len(errs)"`

File: v2/tools/generator/internal/codegen/pipeline/resource_registration_file.go (hunk 2 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/codegen/pipeline/resource_registration_file.go:104]: Pointed out by a linter.

**Pull request merged**